nvme.c

```diff
@@ -5333,6 +5418,571 @@ static int show_registers(int argc, char **argv, struct command *cmd, struct plu
+		return cfg->bpinfo;
+	case NVME_REG_BPRSEL:
+		return cfg->bprsel;
+	case NVME_REG_BPMBL:
+		return cfg->bpmbl;
+	case NVME_REG_CMBMSC:
+		return cfg->cmbmsc;
+	case NVME_REG_CMBSTS:
+		return cfg->cmbsts;
+	case NVME_REG_CMBEBS:
+		return cfg->cmbebs;
+	case NVME_REG_CMBSWTP:
+		return cfg->cmbswtp;
+	case NVME_REG_NSSD:
+		return cfg->nssd;
+	case NVME_REG_CRTO:
+		return cfg->crto;
+	case NVME_REG_PMRCAP:
+		return cfg->pmrcap;
+	case NVME_REG_PMRCTL:
+		return cfg->pmrctl;
+	case NVME_REG_PMRSTS:
+		return cfg->pmrsts;
+	case NVME_REG_PMREBS:
+		return cfg->pmrebs;
+	case NVME_REG_PMRSWTP:
+		return cfg->pmrswtp;
+	case NVME_REG_PMRMSCL:
+		return cfg->pmrmscl;
+	case NVME_REG_PMRMSCU:
+		return cfg->pmrmscu;
+	default:
+		break;
+	}
+
+	return false;
+}
+
+static int get_register_properties(int fd, void **pbar, struct get_reg_config *cfg)
+{
+	int err;
+	int offset = NVME_REG_CRTO;
+	int size = offset + get_reg_size(offset);
+	__u64 value;
+	void *bar = malloc(size);
+	struct nvme_get_property_args args = {
+		.args_size = sizeof(args),
+		.fd = fd,
+		.value = &value,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+	};
+
+	if (!bar) {
+		nvme_show_error("malloc: %s", strerror(errno));
+		return -1;
+	}
+
+	for (offset = NVME_REG_CAP; offset <= NVME_REG_CRTO; offset += get_reg_size(offset)) {
+		if ((cfg->offset != offset && !is_set_reg(cfg, offset)) ||
+		    !nvme_is_fabrics_reg(offset))
+			continue;
+		args.offset = offset;
+		err = nvme_get_property(&args);
+		if (nvme_status_equals(err, NVME_STATUS_TYPE_NVME, NVME_SC_INVALID_FIELD)) {
+			value = -1;
+		} else if (err) {
+			nvme_show_error("get-property: %s", nvme_strerror(errno));
+			free(bar);
+			return err;
+		}
+
+		if (nvme_is_64bit_reg(offset))
+			*(uint64_t *)(bar + offset) = value;
+		else
+			*(uint32_t *)(bar + offset) = value;
+	}
+
+	*pbar = bar;
+
+	return 0;
+}
+
+bool nvme_is_ctrl_reg(int offset)
+{
+	switch (offset) {
+	case NVME_REG_CAP:
+	case NVME_REG_VS:
+	case NVME_REG_INTMS:
+	case NVME_REG_INTMC:
+	case NVME_REG_CC:
+	case NVME_REG_CSTS:
+	case NVME_REG_NSSR:
+	case NVME_REG_AQA:
+	case NVME_REG_ASQ:
+	case NVME_REG_ACQ:
+	case NVME_REG_CMBLOC:
+	case NVME_REG_CMBSZ:
+	case NVME_REG_BPINFO:
+	case NVME_REG_BPRSEL:
+	case NVME_REG_BPMBL:
+	case NVME_REG_CMBMSC:
+	case NVME_REG_CMBSTS:
+	case NVME_REG_CMBEBS:
+	case NVME_REG_CMBSWTP:
+	case NVME_REG_NSSD:
+	case NVME_REG_CRTO:
+	case NVME_REG_PMRCAP:
+	case NVME_REG_PMRCTL:
+	case NVME_REG_PMRSTS:
+	case NVME_REG_PMREBS:
+	case NVME_REG_PMRSWTP:
+	case NVME_REG_PMRMSCL:
+	case NVME_REG_PMRMSCU:
+		return true;
+	default:
+		break;
+	}
+
+	return false;
+}
+
+static bool get_register_offset(void *bar, bool fabrics, struct get_reg_config *cfg,
+				enum nvme_print_flags flags)
+{
+	bool offset_matched = cfg->offset >= 0;
+	int offset;
+
+	if (offset_matched)
+		nvme_show_ctrl_register(bar, fabrics, cfg->offset, flags);
+
+	for (offset = NVME_REG_CAP; offset <= NVME_REG_PMRMSCU; offset += get_reg_size(offset)) {
```

Comment:
If I read this correctly, `get_register_offset` is only used to display one register. So you we know the register which want to display, it's `cfg->offset`, right? If so, we don't need this loop here.

Reply:
No the loop needed to show registers set by cfg parameter specified by the command arguments and it is checked by the function `is_set_reg()`. Sorry for the complecated code.

Reply:
okay, I'll rename this function in this case to `is_reg_selected` which makes it a bit clearer for me.